File: src/0038-countAndSay/countAndSay.cpp
#include <iostream>
#include <string>
using namespace std;



string countAndSay(int n) {

    if (n == 1) return "1";
        
    string s = "1";
    for (int i = 2; i <= n; i++)
    {
        s = countAndSayStr(s);
    }

    return s;

}

string countAndSayStr(string s) {
    string ans;

    int couter = 1;

    int length = s.length();

    for (int i = 1; i< length ; i++)
    {
        if (s[i] == s[i-1]) {
            couter++;
        } else {
            ans = ans + to_string(couter) + s[i-1];
            couter = 1;
        }
    }

    ans = ans + to_string(couter) + s[length-1];


    return ans;


}


int main()
{
    cout << countAndSay(20) << endl;
    return 0;
}

File: src/0038-countAndSay/CountAndSay.swift
// recursive version
func countAndSay(_ n: Int) -> String {
    
    if n == 1 {
        return "1"
    }
    
    let lastStr = countAndSay(n-1);
    var ans = ""
    var cur = lastStr.first
    var count = 0
    for c in lastStr {
        if c == cur {
            count += 1
        } else {
            ans.append("\(count)"+"\(cur ?? Character(""))")
            cur = c
            count = 1
        }
    }
    ans.append("\(count)"+"\(cur ?? Character(""))")
    return ans
}


func countAndSay(_ n: Int) -> String {
    
    if n == 1 {
        return "1"
    }
    
    let lastStr = countAndSay(n-1);
    var ans = ""
    var cur = lastStr.first
    var count = 0
    for c in lastStr {
        if c == cur {
            count += 1
        } else {
            ans.append("\(count)"+"\(cur ?? Character(""))")
            cur = c
            count = 1
        }
    }
    ans.append("\(count)"+"\(cur ?? Character(""))")
    return ans
}

func countAndSayIterator(_ n: Int) -> String {
    var pre = "1"

    for i in 2...n {
        var count = 0
        var tmp = ""
        var current = pre.first
        
        for c in pre {
            if c == current {
                count += 1
            } else {
                tmp.append("\(count)"+"\(current ?? Character(""))")
                current = c
                count = 1
            }
        }
        tmp.append("\(count)"+"\(current ?? Character(""))")
        pre = tmp
    }
    return pre
}


print(countAndSay(30))

print(countAndSayIterator(4))